### PR TITLE
*: collect execution details and output them in slow query log

### PR DESCRIPTION
**Pull request by coocood**
Collect each execution detail in coprocessor request into StatementContext.
Output them in slow query log.

## What have you changed? (mandatory)
Collect each execution detail in coprocessor request into StatementContext.
Output them in slow query log.

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
manual and unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
YES

## Does this PR affect tidb-ansible update? (mandatory)
NO

## Does this PR need to be added to the release notes? (mandatory)
YES

release note:
include execution details in slow query log

### CODE REVIEW

**Comment by coocood** [comment]
@zz-jason @shenli @tiancaiamao PTAL

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -236,3 +238,23 @@ func (sc *StatementContext) ResetForRetry() {
...
+
+// MergeExecDetails merges a single region execution details into self, used to print
+// the information in slow query log.
+func (sc *StatementContext) MergeExecDetails(details *execdetails.ExecDetails) {
+	sc.mu.Lock()
+	sc.mu.execDetails.ProcessTime += details.ProcessTime
+	sc.mu.execDetails.WaitTime += details.WaitTime
+	sc.mu.execDetails.BackoffTime += details.BackoffTime
+	sc.mu.execDetails.RequestCount++
+	sc.mu.Unlock()
```
Comment (lysu) [inline sessionctx/stmtctx/stmtctx.go]: should we need to make execDetails r/w lock-free?

Comment (coocood) [reply sessionctx/stmtctx/stmtctx.go]: No, the lock contention is very low.

File: util/execdetails/execdetails.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	ProcessTime   time.Duration
+	WaitTime      time.Duration
+	BackoffTime   time.Duration
+	RequestCount  int
+	TotalKeys     int64
+	ProcessedKeys int64
+}
+
+// String implements the fmt.Stringer interface.
+func (d ExecDetails) String() string {
```
[the remark this answers is not in the document]
Comment (coocood) [reply util/execdetails/execdetails.go]: This is only called once for a slow query, I think we don't need to worry about the performance.

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply util/execdetails/execdetails.go]: `Formatter` is not necessary, we can use a string builder for it.
https://godoc.org/strings#Builder

File: util/execdetails/execdetails.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		parts = append(parts, fmt.Sprintf("wait_time:%v", d.WaitTime))
+	}
+	if d.BackoffTime > 0 {
+		parts = append(parts, fmt.Sprintf("backoff_time:%v", d.BackoffTime))
+	}
+	if d.RequestCount > 0 {
+		parts = append(parts, fmt.Sprintf("request_count:%d", d.RequestCount))
+	}
+	if d.TotalKeys > 0 {
+		parts = append(parts, fmt.Sprintf("total_keys:%d", d.TotalKeys))
```
Comment (zz-jason) [inline util/execdetails/execdetails.go]: what does "total_keys" and "processed_keys" mean?

File: distsql/select_result.go [hunk 4]
```diff
@@ -159,11 +159,13 @@ func (r *selectResult) getSelectResp() error {
...
 			return terror.ClassTiKV.New(terror.ErrCode(err.Code), err.Msg)
 		}
+		sc := r.ctx.GetSessionVars().StmtCtx
 		for _, warning := range r.selectResp.Warnings {
-			r.ctx.GetSessionVars().StmtCtx.AppendWarning(terror.ClassTiKV.New(terror.ErrCode(warning.Code), warning.Msg))
+			sc.AppendWarning(terror.ClassTiKV.New(terror.ErrCode(warning.Code), warning.Msg))
 		}
 		r.feedback.Update(re.result.GetStartKey(), r.selectResp.OutputCounts)
 		r.partialCount++
+		sc.MergeExecDetails(re.result.GetExecDetails())
```
Comment (zz-jason) [inline distsql/select_result.go]: It's better to print execution details for every coprocessor request.

File: util/execdetails/execdetails.go (hunk 3 above)
Comment (coocood) [reply util/execdetails/execdetails.go]: We can explain them in documents later.

File: distsql/select_result.go (hunk 4 above)
Comment (coocood) [reply distsql/select_result.go]: Sometimes there are hundreds of coprocessor requests, then the SLOW_QUERY log would be unreadable.

Comment (zz-jason) [reply distsql/select_result.go]: sometime there is a data skew in some key ranges, and this data skew can slow down the coprocessor performance and cause other coprocessor request waits for processing. 

if we merge all the coprocessor request for a user query, we can not recognize the data skew.

Comment (coocood) [reply distsql/select_result.go]: How about we maintain a max?

Comment (coocood) [reply distsql/select_result.go]: The main problem this PR solves is to distinguish culprit queries among SLOW queries.
The total time is enough for this purpose.

**Comment by zz-jason** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**